Fix list all access list members/reviews pagination.

mdwn opened the pull request:
Two things were happening that was preventing the cache of access list members/reviews from initializing if the number of members was greater than the default page size of 200:

1. The incorrect page token was being used in the ListAllAccessListMembers call.
2. The pagination key for listing all of these members was incorrect due to the use of sub-prefixes.

Both issues have been corrected. The ListAllAccessListMembers/Reviews calls in the access list service now uses their own custom pagination logic.

changelog: Fix cache init issue with access list members/reviews.

espadolini commented on lib/services/local/access_list_test.go:
```diff
@@ -1353,6 +1354,51 @@ func TestChangingOwnershipModeIsAnError(t *testing.T) {
+		}
+	}
+
+	require.Equal(t, numAccessLists*numAccessListMembersPerAccessList, len(allMembers))
```
If possible I'd actually check that we got back all and only the members that we expected and in the intended order.

espadolini commented on lib/services/local/generic/generic.go:
```diff
@@ -173,7 +173,16 @@ func (s *Service[T]) ListResources(ctx context.Context, pageSize int, pageToken
 	var nextKey string
 	if len(out) > pageSize {
-		nextKey = backend.GetPaginationKey(out[len(out)-1])
+		lastElementIndex := len(out) - 1
+
+		// Add in any subkeys if present.
+		itemKeyWithoutPrefix := string(result.Items[lastElementIndex].Key[len(s.backendPrefix)+2:])
+		lastSeparatorIndex := strings.LastIndex(itemKeyWithoutPrefix, string(backend.Separator))
+
+		nextKey = backend.GetPaginationKey(out[lastElementIndex])
+		if lastSeparatorIndex != -1 {
+			nextKey = strings.Join(append([]string{}, itemKeyWithoutPrefix[:lastSeparatorIndex], nextKey), string(backend.Separator))
+		}
```
Why isn't this basically just `result.Items[len(out)-1].Key[2+len(s.backendPrefix):]`? I'd be wary of changing things like this, since `GetPaginationKey` handles resources with a HostID in a special way.

zmb3 commented on lib/services/local/access_list_test.go:
```diff
@@ -1353,6 +1354,55 @@ func TestChangingOwnershipModeIsAnError(t *testing.T) {
+	// Create several access lists.
+	expectedMembers := make([]*accesslist.AccessListMember, totalMembers)
+	for i := 0; i < numAccessLists; i++ {
+		alName := fmt.Sprintf("%d", i)
```
```suggestion
		alName := strconv.Itoa(i)
```

zmb3 commented on lib/services/local/generic/generic.go:
```diff
@@ -173,7 +173,7 @@ func (s *Service[T]) ListResources(ctx context.Context, pageSize int, pageToken
 	var nextKey string
 	if len(out) > pageSize {
-		nextKey = backend.GetPaginationKey(out[len(out)-1])
+		nextKey = string(result.Items[len(out)-1].Key[2+len(s.backendPrefix):])
```
I have a hard time even telling what this code is trying to do. There's no comment explaining it, nor is there any bounds checking.

mdwn replied:
@rosstimothy Maybe. It depends on whether host IDs are basically handled the same way. If it's: `<host-id>/<name>`, this is functionally the same.

@zmb3 Basically we're trying to ensure that the pagination key has any sub-keys, e.g. the pagination key of
`/access_list_member/access-list-name/member-name`

should be `access-list-name/member-name` and not just `member-name`.

I can add in a comment and bounds checking, but just wanted to verify that his looks okay before I chase it down further.

Commit:
Fix list all access list members pagination.

Two things were happening that was preventing the cache of access list
members from initializing if the number of members was greater than the
default page size of 200:

1. The incorrect page token was being used in the ListAllAccessListMembers
   call.
2. The pagination key for listing all of these members was incorrect due
   to the use of sub-prefixes.

Both issues have been corrected. The ListAllAccessListMembers call in the
access list service now uses its own custom pagination logic.

Commit:
Realized reviews suffer from the same problem.

espadolini commented on lib/services/local/access_list.go:
```diff
@@ -271,10 +273,49 @@ func (a *AccessListService) ListAccessListMembers(ctx context.Context, accessLis
+		// For listing all access list members for all access lists, we want to list members regardless of which
+		// access list they belong to. As a result, we'll use custom pagination logic here to include the appropriate
+		// prefix across pages.
+		lastItemKey := result.Items[len(out)-1].Key
+		prefixOffset := 2 + len(accessListMemberPrefix)
+		if len(lastItemKey) < prefixOffset {
+			return nil, "", trace.BadParameter("unable to calculate next page for listing all access list members")
+		}
+		nextKey = string(lastItemKey[prefixOffset:])
```
Here and in `ListAllAccessListReviews`: the logic should be the same as the one used to pick the key for upsert, it's just a bit hidden on upsert because of the `WithPrefix` service wrapper.
```suggestion
		nextKey = out[pageSize].Spec.AccessList + string(backend.Separator) + out[pageSize].Metadata.Name
```

espadolini replied:
Maybe the generic service could grow an extra optional `func(T) string` to get the intended backend key to append to the prefix and to make pagination keys from, rather than relying on the simple heuristic that prepends a host id if the resource has a `GetHostID() string`?

mdwn replied:
> Can we drop all of the custom listing here if we increase the page size by one and then create the true next key from the last access list member or access list review?

This doesn't work because the generic won't let pages exceed the max size set at initialization.

mdwn replied:
I've adjusted this per our Slack conversation and introduced a `ListResourcesReturnNextResource`. Let me know what you think.

Commit:
Simplify nextKey calc.

Commit:
Use a function that returns the next element and allow the caller to construct their own next key.

Commit:
Use *T instead of making Resource comparable.

espadolini commented on lib/services/local/generic/generic.go:
```diff
@@ -159,26 +170,25 @@ func (s *Service[T]) ListResources(ctx context.Context, pageSize int, pageToken
 	result, err := s.backend.GetRange(ctx, rangeStart, rangeEnd, limit)
 	if err != nil {
-		return nil, "", trace.Wrap(err)
+		return nil, next, trace.Wrap(err)
```
```suggestion
		return nil, nil, trace.Wrap(err)
```

espadolini commented on lib/services/local/generic/generic.go:
```diff
@@ -159,26 +170,25 @@ func (s *Service[T]) ListResources(ctx context.Context, pageSize int, pageToken
 		resource, err := s.unmarshalFunc(item.Value, services.WithRevision(item.Revision), services.WithResourceID(item.ID))
 		if err != nil {
-			return nil, "", trace.Wrap(err)
+			return nil, next, trace.Wrap(err)
```
```suggestion
			return nil, nil, trace.Wrap(err)
```

espadolini commented on lib/services/local/generic/generic.go:
```diff
@@ -159,26 +170,25 @@ func (s *Service[T]) ListResources(ctx context.Context, pageSize int, pageToken
-	var nextKey string
 	if len(out) > pageSize {
-		nextKey = backend.GetPaginationKey(out[len(out)-1])
+		next = &out[len(out)-1]
```
```suggestion
		next = &out[pageSize]
```

Commit:
Cleanup code, use nils instead of named return, eliminate named returns to avoid confusion.
